### ISSUE TITLE
Mark daprd as a sidecar container using the SideCar type in the container's lifecycle spec

### ISSUE THREAD

**Issue by stephaneey** [dapr/dapr#1907]
## In what area(s)?

> /area runtime
 /area operator
> /area placement
> /area docs
> /area test-and-release

## Describe the feature

I noticed that the application container might be ready before the daprd one, especially because of the initial delay added to the daprd container. Since K8s 1.18, it is possible to mark a container as a SideCar so that K8s makes sure the SideCar container starts before others. By reverse-engineering an injected Pod, I could see that this is not used today by Dapr. It would certainly be a breaking change but since you're still in Alpha mode, that might be the right time.

To illustrate this, if I write the follwowing busybox:

`apiVersion: v1
kind: Pod
metadata:
  name: busyboxwithdapr
  namespace: azug
  labels:
    app: busyboxwithdapr
    aadpodidbinding: busyboxidentity
  annotations:        
        dapr.io/enabled: "true"        
        dapr.io/log-level: "debug"
spec:
  containers:
  - image: yauritux/busybox-curl
    command: [ "/bin/sh", "-c", "--" ]
    args: [ "while true; do echo busybox;curl --request GET http://127.0.0.1:3500/v1.0/secrets/mysecretstore/storageKey; sleep 31; done;" ]
    imagePullPolicy: IfNotPresent
    name: busybox
  restartPolicy: Always`

that immediately calls a secret store, I initially get errors back from the curl command because daprd isn't ready yet. 

`busybox
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (7) Failed connect to 127.0.0.1:3500; Connection refused
busybox
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   105  100   105    0     0    172      0 --:--:-- --:--:-- --:--:--   172
{"storageKey":"o2q620...=="}busybox
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed`

This could cause the application container (not in this case) not to restart properly in case of poor exception handling and with a restartPolicy of Never.

**Comment by yaron2** [comment]
The sidecar mode is extremely interesting, I was under the impression this is still being designed.

**Comment by yaron2** [comment]
So this lands in 1.19, but we will enable that. I don't see why this is a breaking change for us?

**Comment by stephaneey** [comment]
If you make use of this feature on older clusters (say someone upgrades dapr on an older cluster), I suppose this will be considered invalid...or I might be missing something?

**Comment by yaron2** [comment]
> If you make use of this feature on older clusters (say someone upgrades dapr on an older cluster), I suppose this will be considered invalid...or I might be missing something?

It will not pass validation for < 1.19, but we can enable it on demand using an annotation. This is good info, thanks!

**Comment by artursouza** [comment]
@fanfanbj Do you think this can be done for 1.7? Release date will probably be within 45 days.

**Comment by fanfanbj** [comment]
@artursouza I made a little to investigate sidecar container, it hasn't release in kubernetes. There are some solutions to solve this kind of problems. So this problem maybe delay to close.

**Comment by fanfanbj** [comment]
@yaron2 @artursouza I found that sidecar container type feature is still undecided in kubernetes. Dapr provides a way to determine it’s health using an HTTP /healthz endpoint：GET http://localhost:<daprPort>/v1.0/healthz. App can call this API at prestart in container's lifecyle to determine sidecar's start status. Can we use this way as a solution? and we can log this into dapr document until the kubernetes sidecar feature released. And this way can be used as a common solution across all kubernetes version.

**Comment by sicoyle** [comment]
> I think we should close this issue. The latest I read on this from the relevant Kubernetes SIGs seems to have abandoned the idea of marking containers as sidecars.

Can this be closed?

**Comment by yaron2** [comment]
No, sidecar containers are now a first class citizen in K8s: https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/.

**Comment by MeeraTresa** [comment]
I am interested in working on this @yaron2

**Comment by sicoyle** [comment]
@MeeraTresa lol oops

Can you comment:
```
/assign
```

so it'll assign you

**Comment by sicoyle** [comment]
@kendallroden Let's leave this in the needs review column. Sidecar setting in k8s is in alpha stage, so best to work on other things atm.

**Comment by PierTrailslight** [comment]
It would be great if dapr sidecars were implemented as k8s sidecar https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/, now that the feature is stable (k8s 1.33).

**Comment by MyMirelHub** [comment]
> It would be great if dapr sidecars were implemented as k8s sidecar https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/, now that the feature is stable (k8s 1.33).

Raised a feature request here for native sidecars https://github.com/dapr/dapr/issues/9422
